### ISSUE TITLE
Overall Model Performance Over Time

### ISSUE THREAD

**Issue by carterkoehler** [automl/auto-sklearn#337]
I'm running some tests with different configurations of auto-sklearn, and I'd like to get some graphs for visualizing how the performance of the overall model progresses over time, somewhat similar to Figure 4 in _Efficient and Robust Automated Machine Learning_. Right now I'm using `cv_results_`, but, as far as I can tell, that only contains data on the individual runs performed within the Bayesian optimization (which ends up being very messy), but I would like to know how the total ensemble is performing at that point/how auto-sklearn would theoretically perform if training was stopped there.

Is there any support for what I'm trying to do, or would I have to do that manually? Alternatively, what was done to produce the aforementioned figures?

**Comment by mfeurer** [comment]
Figure 4 does not show ensemble performance, but shows how the single best model changes over time. For that, whenever a new configuration is evaluated, we check whether it's performance on the validation set improves over the previous best configuration, and if so, we plot the test performance of that configuration. Figure 3 on the other hand shows the ensemble performance over time.

In order to get data to reproduce Figure 4, you can check and modify this file: https://github.com/automl/auto-sklearn/blob/master/scripts/run_auto-sklearn_for_metadata_generation.py

If you want to reproduce Figure 4, I can look for a script to get the ensemble performance at different time steps if you want.

**Comment by carterkoehler** [comment]
Thank you, that was very helpful. I would also like to know if autosklearn logs data to produce a graph similar to Figure 3, or was that done with a separate script.

**Comment by mfeurer** [comment]
Yes and no. Auto-sklearn stores the predictions of each model in order to build an ensemble out of them. With a separate script, an ensemble was built for each time step, checking whether the validation performance is greater than the validation performance of the previous ensemble, and if yes, storing the test performance. If you would like to reproduce such figures, I can search the script.

**Comment by carterkoehler** [comment]
Sure, that would be great.

**Comment by mfeurer** [comment]
Check out the script I pasted below. It does not support choosing a metric, but should do the job.

```
from argparse import ArgumentParser
import csv
import glob
import os
import time

import numpy as np
import sklearn.externals.joblib as joblib

from autosklearn.ensembles.ensemble_selection import EnsembleSelection
from autosklearn.metrics import roc_auc, balanced_accuracy, f1_macro, \
    pac_score, r2
from autosklearn.util.backend import create

metric_name_to_function = {'auc_metric': roc_auc,
                           'bac_metric': balanced_accuracy,
                           'f1_metric': f1_macro,
                           'pac_metric': pac_score,
                           'r2_metric': r2,
                           'roc_auc': roc_auc}


def _load_file(f):
    split = f.split('_')
    as_seed = int(split[-2])
    ta_seed = int(split[-1].split('.')[0])
    np_array = np.load(f)
    return np_array, (as_seed, ta_seed), os.path.getmtime(f)


def read_files(directory, seed=None, n_jobs=1):

    seed_pattern = '*' if seed is None else str(seed)
    glob_pattern = os.path.join(directory, "predictions_*_%s_*.npy" %
                                seed_pattern)
    files = sorted(glob.glob(glob_pattern))
    files = joblib.Parallel(n_jobs=n_jobs, verbose=10)(
        joblib.delayed(_load_file) (f=f) for f in files)

    return files


def main(input_directories, output_file, seed, ensemble_size, n_jobs=1):
    seed = None if seed is None or seed < 0 else int(seed)
    if isinstance(input_directories, str):
        input_directories = [input_directories]

    validation_files = []
    test_files = []
    starttime = np.inf
    for input_directory in input_directories:
        print('Loading files from input directory:', input_directory)
        validation_files_ = read_files(
            os.path.join(input_directory, '.auto-sklearn/predictions_ensemble'),
            n_jobs=n_jobs)
        validation_files.extend(validation_files_)
        test_files_ = read_files(
            os.path.join(input_directory, '.auto-sklearn/predictions_test'),
            n_jobs=n_jobs)
        test_files.extend(test_files_)

        assert len(validation_files_) > 0
        assert len(validation_files_) == len(test_files_)

        print('Loaded %d files!' % len(validation_files_))

        seed_pattern = '*' if seed is None else str(seed)
        glob_pattern = os.path.join(input_directory,
                                    ".auto-sklearn",
                                    "start_time_%s" % seed_pattern)
        start_time_files = glob.glob(glob_pattern)

        for start_time_file in start_time_files:
            with open(start_time_file, "r") as fh:
                starttime_candidate = float(fh.read())
                if starttime_candidate < starttime:
                    starttime = starttime_candidate

        del validation_files_
        del test_files_

    validation_files.sort(key=lambda t: t[-1])

    keys_to_test_files = {test_file[1]: test_file
                          for test_file in test_files}
    # Resort such that both files have the same order
    test_files = [keys_to_test_files[validation_file[1]]
                  for validation_file in validation_files]

    assert [validation_file[1] for validation_file in validation_files] == [
        test_file[1] for test_file in test_files]

    losses = []
    top_models_at_step = dict()

    backend = create(input_directory, input_directory,
                     delete_tmp_folder_after_terminate=False,
                     delete_output_folder_after_terminate=False)
    valid_labels = backend.load_targets_ensemble()
    D = backend.load_datamanager()

    dataset_name = D.name
    if dataset_name == "alexis":
        score = D.info['metric']
    elif dataset_name == "yolanda":
        score = D.info['metric']
    elif dataset_name == "tania":
        score = D.info['metric']
    elif dataset_name == "adult":
        score = D.info['metric']
    elif dataset_name == "dorothea":
        score = D.info['metric']
    elif dataset_name == "digits":
        score = D.info['metric']
    elif dataset_name == "cadata":
        score = D.info['metric']
    elif dataset_name == "newsgroups":
        score = D.info['metric']
    else:
        score = balanced_accuracy

    for i in range(len(validation_files)):
        loss = 1 - score(valid_labels, validation_files[i][0])
        losses.append(loss)
        sorted_losses = np.argsort(losses)[:200]
        top_models_at_step[i] = sorted_losses

    models_to_remove = set(list(range(len(validation_files))))
    for top_models_at_i in top_models_at_step:
        for top_model in top_models_at_step[top_models_at_i]:
            if top_model in models_to_remove:
                models_to_remove.remove(top_model)

    print("Removing the following %d models from the library: %s"
          % (len(models_to_remove), models_to_remove))
    for model_id in models_to_remove:
        validation_files[model_id] = None
        test_files[model_id] = None

    print('Starting ensemble building!')
    output = joblib.Parallel(n_jobs=n_jobs, verbose=20)(
        joblib.delayed(evaluate)(input_directory=input_directories[0],
                                 validation_files=[validation_files[j] for
                                                   j in range(len(validation_files))
                                                   if j in top_models_at_step[i]],
                                 test_files=[test_files[j] for
                                             j in range(len(test_files))
                                             if j in top_models_at_step[i]],
                                 ensemble_size=ensemble_size)
                                 for i in range(len(test_files)))

    # Create output csv file
    file_output_seed = -1 if seed is None else seed
    with open(output_file, "w") as csv_file:
        fieldnames = ['Time', 'Training (Empirical) Performance',
                      'Test Set Performance', 'AC Overhead Time',
                      'Validation Configuration ID']
        csv_writer = csv.DictWriter(csv_file, fieldnames=fieldnames)
        csv_writer.writeheader()

        csv_writer.writerow({'Time': 0,
                             'Training (Empirical) Performance': 1.0,
                             'Test Set Performance': 1.0,
                             'AC Overhead Time': 0,
                             'Validation Configuration ID': 0})

        for i, o in enumerate(output):
            csv_writer.writerow({'Time': o['ensemble_time'] +
                                         o['time_function_evaluation'] -
                                         starttime,
                                 'Training (Empirical) Performance':
                                     o['ensemble_error'],
                                 'Test Set Performance': o['ensemble_test_error'],
                                 'AC Overhead Time': 0,
                                 'Validation Configuration ID': i})


def evaluate(input_directory, validation_files, test_files, ensemble_size=50):

    backend = create(input_directory, input_directory,
                     delete_tmp_folder_after_terminate=False,
                     delete_output_folder_after_terminate=False)

    valid_labels = backend.load_targets_ensemble()
    D = backend.load_datamanager()
    test_labels = D.data["Y_test"]

    dataset_name = D.name
    score = balanced_accuracy

    # Read the modification time of the predictions file and compute the interval to the first prediction file.
    # This interval will be add to the time we needed to build the ensemble
    time_function_evaluation = validation_files[-1][-1]

    # Build the ensemble
    start = time.time()
    ensemble_selection = EnsembleSelection(ensemble_size=ensemble_size,
                                           task_type=D.info['task'],
                                           metric=score)

    validation_predictions = np.array([v[0] for v in validation_files])
    test_predictions = np.array([t[0] for t in test_files])
    ensemble_selection.fit(validation_predictions, valid_labels, identifiers=None)
    y_hat_ensemble = ensemble_selection.predict(np.array(validation_predictions))
    y_hat_test = ensemble_selection.predict(np.array(test_predictions))

    # Compute validation error
    ensemble_error = 1 - score(valid_labels, y_hat_ensemble)

    # Compute test error
    ensemble_test_error = 1 - score(test_labels, y_hat_test)

    ensemble_time = time.time() - start

    rval = {'ensemble_time': ensemble_time,
            'time_function_evaluation': time_function_evaluation,
            'ensemble_error': ensemble_error,
            'ensemble_test_error': ensemble_test_error}

    return rval


if __name__ == '__main__':
    parser = ArgumentParser()
    parser.add_argument('--input-directory', type=str, required=True, nargs='+')
    parser.add_argument('-s', '--seed', type=int)
    parser.add_argument("--output-file", type=str, required=True)
    parser.add_argument("--ensemble-size", type=int, default=50)
    parser.add_argument("--n-jobs", type=int, default=1)
    args = parser.parse_args()

    input_directory = args.input_directory
    output_file = args.output_file
    seed = args.seed
    ensemble_size = args.ensemble_size
    n_jobs = args.n_jobs

    main(input_directory, output_file, seed, ensemble_size, n_jobs)
```

**Comment by mfeurer** [comment]
Closing this because the answer is given. Please reopen if necessary.

**Comment by herilalaina** [comment]
Hello Matthias,

Thank you for this great contribution. We are also willing to reproduce figure 3 in _Efficient and Robust Automated Machine Learning_. But we have issues when running your last script (overall ensemble performance over time). It assumed that the directory `.auto-sklearn/predictions_test` exists which is not the case.

Could you provide more information on how can we get this directory? For now, we ran [this script](https://github.com/automl/auto-sklearn/blob/master/scripts/run_auto-sklearn_for_metadata_generation.py) with slight modification on Auto-Sklearn configuration:
```python
automl_arguments = {
    'time_left_for_this_task': 3600, 
    'per_run_time_limit': 360,
    'initial_configurations_via_metalearning': 0,
    'ensemble_size': 50,
    'ensemble_nbest': 50,
    'seed': seed,
    'ml_memory_limit': 3072,
    'resampling_strategy': 'cv',
    'resampling_strategy_arguments': {'folds': 10},
    'delete_tmp_folder_after_terminate': False,
    'delete_output_folder_after_terminate': False,
    'tmp_folder': tmp_dir,
    'output_folder': output_dir,
    'disable_evaluator_output': False,
    "shared_mode": True,
}
``` 
Another question is about the meta-learning related curve. How much configuration (value of `initial_configurations_via_metalearning`) is considered for the experiments? In order to implement the Leave-One-Dataset-Out strategy, how can we (temporarily) drop the meta-information of the current dataset? 

Thank you.
